src/qibocal/auto/transpile.py

```diff
@@ -97,19 +98,65 @@ def execute_transpiled_circuit(
         backend,
         transpiler,
     )[0]
-
     return transpiled_circ, backend.execute_circuit(
         transpiled_circ, initial_state=initial_state, nshots=nshots
     )
 
 
+def get_natives(platform):
+    """
+    Return the list of native gates defined in the `platform`.
+    This function assumes the native gates to be the same for each
+    qubit and pair.
+    """
+    pairs = list(platform.pairs.values())[0]
+    qubit = list(platform.qubits.values())[0]
+    two_qubit_natives = list(pairs.native_gates.raw)
+    single_qubit_natives = list(qubit.native_gates.raw)
+    # Solve Qibo-Qibolab mismatch
+    single_qubit_natives.append("RZ")
+    single_qubit_natives.append("Z")
+    single_qubit_natives.remove("RX12")
+    replacements = {
+        "RX": "GPI2",
+        "MZ": "M",
+    }
+    new_single_natives = [replacements.get(i, i) for i in single_qubit_natives]
+    natives = new_single_natives + two_qubit_natives
+    return natives
+
+
+def set_compiler(backend, natives):
+    """
+    Set the compiler to execute the native gates defined by the platform.
+    """
+    compiler = backend.compiler
+    for native in natives:
+        gate = getattr(gates, native)
+        if gate not in compiler.rules:
+
+            def rule(qubits_ids, platform, parameters=None):
+                if len(qubits_ids[1]) == 1:
+                    native_gate = platform.qubits[tuple(qubits_ids[1])].native_gates
+                else:
+                    native_gate = platform.pairs[tuple(qubits_ids[1])].native_gates
+                pulses = getattr(native_gate, native).pulses
+                return PulseSequence(pulses), {}
+
+            backend.compiler[gate] = rule
```

Comment:
Also, usually it is better to avoid mutating objects, especially nested ones.
I could lift the suggestion by one level, and suggest creating the entire backend (here just returning the compiler, instead of modifying the backend). But let's limit to the compiler itself.

You could create an empty dictionary `rules`, and here do
```suggestion
            rules[gate] = rule
```
eventually setting the compiler out of the loops as:
```py
    backend.compiler = Compiler(rules=rules)
```
(in principle, you can avoid even modifying the dictionary `rules`, by defining it immediately through comprehension - but I get this is a bit unfamiliar).